### ISSUE TITLE
Modifier for bitflag enums

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#42521]
Consider the following code

```ts
export enum ScriptKind {
  Unknown = 0,
  JS = 1,
  JSX = 2,
  TS = 3,
  TSX = 4,
  External = 5,
  JSON = 6,
}
```

If a user tries to pass along the expression `ScriptKind.TS | ScriptKind.TSX` to a parameter that expects a `ScriptKind`, then will end up constructing a value equal to `ScriptKind.JSON`.

Really, the user should have been told that this enum wasn't intended to be used in bitwise operations! But unfortunately, TypeScript doesn't catch this because you can assign any number to any numeric enum.

One could imagine a mode that prevents both the construction and the assignment for certain enums. In this mode, an enum has to be declared with the `bitflags` modifier.

```
bitflags enum SymbolFlags {
    // ...
}
```

Alternatively, this could be something more akin to `/** @deprecated */` or `/** @internal */`.

```
/// @bitflags
enum SymbolFlags {
    // ...
}
```

This is in some part inspired by Rust's [`bitflags`](https://docs.rs/bitflags/1.2.1/bitflags/) macro.

**Comment by MartinJohns** [comment]
This sounds like a duplicate of #21598.

**Comment by DanielRosenwasser** [comment]
It is pretty similar! The difference is that this modifier is really just a hint for some sort of strictness flag. It has no runtime semantics.

**Comment by ShuiRuTian** [comment]
I like this! I would try to give a prototype.
